Refactor to use display entities

kmcginnes opened the pull request:
<!--
Please read the [Code of Conduct](https://github.com/aws/graph-explorer/blob/main/CODE_OF_CONDUCT.md) and the [Contributing Guidelines](https://github.com/aws/graph-explorer/blob/main/CONTRIBUTING.md) before opening a pull request.
-->

## Description

Nearly all views that show details of a node, edge, or type would do the exact same transformations at the view level. This change abstracts all of that logic to a single place, ensuring the logic is consistent and making it easier to modify the behavior across the app.

The next step of this refactoring will be to abstract the neighbor count logic to the same `DisplayVertex`. And after that we will have a good foundation for loading a node/edge from a simple ID, which is foundational to URL sharing.

## Validation

- Smoke test all areas of the app

## Related Issues

<!--
  Link to the issue that is fixed by this PR (if there is one)
  e.g. Fixes #1234, Addresses #1234, Related to #1234, etc.
-->

- Resolves #677
- Affects #122 

### Check List

<!--
  ATTENTION
  Please follow this check list to ensure that you've followed all items before opening this PR
  You can check the items by adding an `x` between the brackets, like this: `[x]`
-->

- [x] I confirm that my contribution is made under the terms of the Apache 2.0
      license.
- [x] I have run `pnpm checks` to ensure code compiles and meets standards.
- [x] I have run `pnpm test` to check if all tests are passing.
- [x] I have covered new added functionality with unit tests if necessary.
- [x] I have added an entry in the `Changelog.md`.

kmcginnes commented on packages/graph-explorer/src/core/StateProvider/displayAttribute.ts:
```diff
@@ -0,0 +1,51 @@
+  typeAttributes: AttributeConfig[],
+  textTransform: (text?: string) => string
+): DisplayAttribute[] {
+  const nodeAttributeNames = Object.keys(entity.attributes);
```
Update the name

kmcginnes commented on packages/graph-explorer/src/hooks/useDisplayNames.ts:
Double check that edge's properly support the display name selection when customized

kmcginnes replied:
The edge display name logic is in a weird place. I'm going to pull it up like I did with the vertex logic.

andreachild commented on packages/graph-explorer/src/core/ConfigurationProvider/useConfiguration.ts:
```diff
@@ -1,137 +1,125 @@
-    if (!configuration) {
-      return;
-    }
+      const vertexTypesMap = new Map(
```
Nit: It seems this map is only used to obtain unique vertex types via the keys of the map - would it be sufficient to use a Set instead of a Map since the map values are not used?

kmcginnes replied:
At this point, yes. I missed this during my refactoring to remove the functions. Thanks! I'll update.

andreachild commented on packages/graph-explorer/src/core/StateProvider/displayAttribute.ts:
```diff
@@ -0,0 +1,52 @@
+import { Vertex, Edge } from "@/@types/entities";
+import { AttributeConfig } from "../ConfigurationProvider";
```
Nit: IntelliJ is saying this could be `import { AttributeConfig } from "@/core";`

andreachild commented on packages/graph-explorer/src/core/StateProvider/displayEdge.ts:
```diff
@@ -0,0 +1,154 @@
+}
+
+/** Finds the `DisplayEdge` instance for a give `EdgeId` within the edges added to the graph canvas. */
+export function useDisplayEdge(id: EdgeId) {
```
This function seems unused - is it meant to be used in a future PR?

kmcginnes replied:
Yes, but it is super easy to re-add. Removing for now.

andreachild commented on packages/graph-explorer/src/core/StateProvider/displayEdge.ts:
```diff
@@ -0,0 +1,154 @@
+/** Maps all `Edge` instances which are selected in the graph canvas to `DisplayEdge` instances. */
+export function useSelectedDisplayEdges() {
+  const selectedIds = useRecoilValue(edgesSelectedIdsAtom);
+  const displayEdges = useDisplayEdgesInCanvas();
```
Nit: if there is only one selected edge but many displayed edges, is this doing unnecessary work to iterate through all of the displayed edges when only the selected one needs to be mapped?

kmcginnes replied:
Fair point. Technically, all the display nodes and edges will be created already, but this code doesn't need to force that behavior.

andreachild commented on packages/graph-explorer/src/modules/EntitiesTabular/components/NodesTabular.tsx:
```diff
@@ -70,40 +62,33 @@ const NodesTabular = forwardRef<TabularInstance<ToggleVertex>, any>(
-          },
+          filter: (rows, _columnIds, filterValue) =>
+            rows.filter(row =>
+              row.original.displayTypes
```
Can `original.displayTypes` have multiple comma separated values here? And if so will it be able to be filtered correctly using `match(filterValue.toLowerCase())`?

andreachild commented on packages/graph-explorer/src/modules/EntityDetails/EntityDetails.tsx:
```diff
@@ -14,47 +13,22 @@ import {
-  const selectedEdgesIds = useRecoilValue(edgesSelectedIdsAtom);
   const [userLayout, setUserLayout] = useRecoilState(userLayoutAtom);
+  const selectedNodes = useSelectedDisplayVertices();
+  const selectedNode = selectedNodes[0];
```
What would happen if no nodes were selected?

kmcginnes commented on packages/graph-explorer/src/core/StateProvider/displayEdge.ts:
```diff
@@ -0,0 +1,154 @@
+}
+
+/** Maps a `Edge` instance to a `DisplayEdge` instance using the schema and any user preferences. */
+export function useDisplayEdgeFromEdge(edge: Edge) {
```
Yes, but it is super easy to re-add. Removing for now.

kmcginnes commented:
I've pushed a series of changes:

- Changed the `defaultVertexTypeConfig` and `defaultEdgeTypeConfig` to not have a display label in order to allow the display logic to fill in the display label as needed
- Added test for sparql case of `displayId` and `displayLabel`
- Updated some of the imports to be the aliased import
- Made `useSelectedDisplay*` hooks more efficient
- Updated styles of node and edge properties in the details sidebar
- Added default properties for edges:
  - Edge type
  - Source vertex id & type
  - Target vertex id & type
- Move edge display name logic to the `DisplayEdge` type rather than `useGraphStyles()`

kmcginnes commented:
Fixed tests and added some more

andreachild commented on packages/graph-explorer/src/modules/EntityDetails/EdgeDetail.tsx:
```diff
@@ -37,7 +37,7 @@ const EdgeDetail = ({ edge }: EdgeDetailProps) => {
         <div className={"content"}>
           <div className={"title"}>{edge.displayTypes}</div>
-          {edge.displayId && <div>{edge.displayId}</div>}
+          {edge.hasUniqueId ? <div>{edge.displayId}</div> : null}
```
Curious why this check is preferred to the original check for `displayId`? Is there still value in setting the `displayId` on the `EdgeDetail` for SPARQL if it's not shown?

kmcginnes replied:
I honestly don't remember why I made that change. If I had to guess it was to make sure no properties were optional on the model. But that's not a good enough reason.

I reverted the change.

kmcginnes replied:
Actually, I take it all back. After reverting the change and seeing some errors I now remember why. It is about optionality, but specifically around the logic to get the "display name" value. Since `displayId` is an option on non-sparql databases if it was optional it would require display name to be optional, which I do not want.

andreachild commented:
LGTM 👍
